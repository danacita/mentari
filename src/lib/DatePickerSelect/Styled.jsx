import styled, { css } from 'styled-components';
import Color from '../Styles/bases/Color';

const SelectWrapper = styled.div`
  position: relative;
  width: 100%;
  select {
    position: relative;
    background-color: white;
    padding: 8px 16px;
    border: 1px solid ${Color.grey191};
    line-height: 1.9;
    -moz-appearance: none;
    -webkit-appearance: none;
    appearance: none;
    &.dc-date {
      width: 25%;
    }
    &.dc-month {
      width: calc(50% - 32px);
      margin: 0 16px;
    }
    &.dc-year {
      width: 25%;
    }
    background-image: linear-gradient(45deg, transparent 50%, gray 50%),
      linear-gradient(135deg, gray 50%, transparent 50%),
      linear-gradient(to right, transparent, transparent);
    background-position: calc(100% - 20px) calc(1.2em + 2px), calc(100% - 15px) calc(1.2em + 2px),
      calc(100% - 2.5em) 0.5em;
    background-size: 8px 5px, 6px 5px, 1px 1.5em;
    background-repeat: no-repeat;
  }

  ${props =>
    props.touched && props.error
      ? css`
          select {
            border: 1px solid ${Color.red};
          }
        `
      : null};

  ${props =>
    props.disabled
      ? css`
          select {
            background-color: ${Color.white};
            cursor: not-allowed;
            border: 1px solid ${Color.grey239};
          }
        `
      : null};
`;

export default SelectWrapper;
