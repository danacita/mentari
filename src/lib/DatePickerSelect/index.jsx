/* eslint-disable no-lone-blocks */
/* eslint-disable no-unused-expressions */
/* eslint-disable no-nested-ternary */
/* eslint-disable react/no-did-update-set-state */
/* eslint-disable no-plusplus */
/* eslint-disable class-methods-use-this */
import React, { Component, Fragment } from 'react';
import moment from 'moment';
import 'moment/locale/id';
import 'moment/locale/tl-ph';
import PropTypes from 'prop-types';
import { Margin } from '../Styles/utils';
import Color from '../Styles/bases/Color';
import SelectWrapper from './Styled';

const propTypes = {
  id: PropTypes.string,
  testid: PropTypes.string,
  label: PropTypes.string,
  defaultPickerValue: PropTypes.string,
  language: PropTypes.string,
  disabled: PropTypes.bool,
  yearMaxOption: PropTypes.number,
  yearMinOption: PropTypes.number,
  withRedux: PropTypes.bool,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    value: PropTypes.string,
    onChange: PropTypes.func
  }),
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool
  })
};

const defaultProps = {
  id: '',
  testid: '',
  label: '',
  defaultPickerValue: '',
  disabled: false,
  yearMaxOption: 0,
  yearMinOption: 0,
  withRedux: true,
  input: {},
  meta: {},
  language: ''
};

class DatePickerSelect extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectDate: '',
      selectMonth: '',
      selectYear: '',
      valueHidden: ''
    };
    this.renderDateOption = this.renderDateOption.bind(this);
    this.renderMonthOption = this.renderMonthOption.bind(this);
    this.renderYearOption = this.renderYearOption.bind(this);
    this.renderPlaceholder = this.renderPlaceholder.bind(this);
    this.handleChangeDate = this.handleChangeDate.bind(this);
  }

  componentDidUpdate(prevProps) {
    const { disabled } = this.props;
    if (prevProps.disabled !== disabled) {
      this.setState({
        selectDate: '',
        selectMonth: '',
        selectYear: ''
      });
    }
  }

  handleChangeDate(e) {
    const { selectDate, selectMonth, selectYear } = this.state;
    const { input, defaultPickerValue, withRedux } = this.props;
    this.setState(
      {
        [e.target.name]: e.target.value
      },
      () => {
        {
          withRedux
            ? input.onChange(
                `${selectYear || moment(defaultPickerValue).format('YYYY')}-${selectMonth ||
                  moment(defaultPickerValue).format('MM')}-${selectDate ||
                  moment(defaultPickerValue).format('DD')}`
              )
            : '';
        }
      }
    );
  }

  renderDateOption(date) {
    const options = [];
    for (let i = 1; i < date + 1; i += 1) {
      options.push(
        <option value={i < 10 ? `0${i}` : i} key={i}>
          {i < 10 ? `0${i}` : i}
        </option>
      );
    }
    return options;
  }

  renderMonthOption(date) {
    const { language } = this.props;
    const options = [];
    for (let i = 1; i < date + 1; i += 1) {
      options.push(
        <option value={i < 10 ? `0${i}` : i} key={i}>
          {moment()
            .month(i - 1)
            .locale(language ? `${language}` : 'id')
            .format('MMMM')}
        </option>
      );
    }
    return options;
  }

  renderPlaceholder(language) {
    switch (language) {
      case 'en':
        return ['Date', 'Month', 'Year'];
      case 'tl-ph':
        return ['Araw', 'Buwan', 'Taon'];
      default:
        return ['Tanggal', 'Bulan', 'Tahun'];
    }
  }

  renderYearOption(start, end) {
    const options = [];
    for (let i = start; i > end; i--) {
      options.push(
        <option value={i} key={i}>
          {i}
        </option>
      );
    }
    return options;
  }

  render() {
    const { selectDate, selectMonth, selectYear, valueHidden } = this.state;
    const {
      id,
      label,
      testid,
      defaultPickerValue,
      input,
      yearMaxOption,
      yearMinOption,
      disabled,
      language,
      withRedux,
      meta: { error, touched }
    } = this.props;

    const placeholder = this.renderPlaceholder(language);
    return (
      <Fragment>
        {withRedux ? (
          <SelectWrapper error={error} testid={testid} disabled={disabled}>
            <input
              type="hidden"
              id={id}
              name={input.name}
              value={valueHidden || input.value ? input.value : defaultPickerValue}
              onChange={() => {}}
              onBlur={input.onBlur}
            />
            <p className="label-select">{label}</p>
            <select
              name="selectDate"
              value={disabled ? selectDate : selectDate || moment(defaultPickerValue).format('DD')}
              onChange={this.handleChangeDate}
              onBlur={this.handleChangeDate}
              className="dc-date"
              disabled={disabled}
            >
              <option value="">{placeholder[0]}</option>
              {this.renderDateOption(31)}
            </select>
            <select
              name="selectMonth"
              value={
                disabled ? selectMonth : selectMonth || moment(defaultPickerValue).format('MM')
              }
              onChange={this.handleChangeDate}
              onBlur={this.handleChangeDate}
              className="dc-month"
              disabled={disabled}
            >
              <option value="">{placeholder[1]}</option>
              {this.renderMonthOption(12)}
            </select>
            <select
              name="selectYear"
              value={
                disabled ? selectYear : selectYear || moment(defaultPickerValue).format('YYYY')
              }
              onChange={this.handleChangeDate}
              onBlur={this.handleChangeDate}
              className="dc-year"
              disabled={disabled}
            >
              <option value="">{placeholder[2]}</option>
              {this.renderYearOption(
                new Date().getFullYear() + yearMaxOption,
                new Date().getFullYear() - (1 + yearMinOption)
              )}
            </select>
            {touched && error && (
              <Margin top={8}>
                <span style={{ color: Color.red }}>{error}</span>
              </Margin>
            )}
          </SelectWrapper>
        ) : (
          <SelectWrapper testid={testid} disabled={disabled}>
            <input
              type="hidden"
              id={id}
              name={input.name}
              value={valueHidden ? '' : defaultPickerValue}
              onChange={() => {}}
            />
            <p className="label-select">{label}</p>
            <select
              name="selectDate"
              value={disabled ? selectDate : selectDate || moment(defaultPickerValue).format('DD')}
              onChange={this.handleChangeDate}
              className="dc-date"
              disabled={disabled}
            >
              <option value="">{placeholder[0]}</option>
              {this.renderDateOption(31)}
            </select>
            <select
              name="selectMonth"
              value={
                disabled ? selectMonth : selectMonth || moment(defaultPickerValue).format('MM')
              }
              onChange={this.handleChangeDate}
              className="dc-month"
              disabled={disabled}
            >
              <option value="">{placeholder[1]}</option>
              {this.renderMonthOption(12)}
            </select>
            <select
              name="selectYear"
              value={
                disabled ? selectYear : selectYear || moment(defaultPickerValue).format('YYYY')
              }
              onChange={this.handleChangeDate}
              className="dc-year"
              disabled={disabled}
            >
              <option value="">{placeholder[2]}</option>
              {this.renderYearOption(
                new Date().getFullYear() + yearMaxOption,
                new Date().getFullYear() - (1 + yearMinOption)
              )}
            </select>
          </SelectWrapper>
        )}
      </Fragment>
    );
  }
}

DatePickerSelect.propTypes = propTypes;
DatePickerSelect.defaultProps = defaultProps;

export default DatePickerSelect;
