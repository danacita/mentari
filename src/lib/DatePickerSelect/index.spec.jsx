import React from 'react';
import renderer from 'react-test-renderer';

import DatePickerSelect from '.';

describe('DatePickerSelect', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer.create(<DatePickerSelect input={input} meta={meta} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
