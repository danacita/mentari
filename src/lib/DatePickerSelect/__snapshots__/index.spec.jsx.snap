// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerSelect renders correctly 1`] = `
<div
  className="sc-htpNat kYYGhy"
  disabled={false}
>
  <input
    id=""
    onChange={[Function]}
    type="hidden"
    value="Value"
  />
  <p
    className="label-select"
  >
    
  </p>
  <select
    className="dc-date"
    disabled={false}
    name="selectDate"
    onBlur={[Function]}
    onChange={[Function]}
    value="Invalid date"
  >
    <option
      value=""
    >
      Tanggal
    </option>
    <option
      value="01"
    >
      01
    </option>
    <option
      value="02"
    >
      02
    </option>
    <option
      value="03"
    >
      03
    </option>
    <option
      value="04"
    >
      04
    </option>
    <option
      value="05"
    >
      05
    </option>
    <option
      value="06"
    >
      06
    </option>
    <option
      value="07"
    >
      07
    </option>
    <option
      value="08"
    >
      08
    </option>
    <option
      value="09"
    >
      09
    </option>
    <option
      value={10}
    >
      10
    </option>
    <option
      value={11}
    >
      11
    </option>
    <option
      value={12}
    >
      12
    </option>
    <option
      value={13}
    >
      13
    </option>
    <option
      value={14}
    >
      14
    </option>
    <option
      value={15}
    >
      15
    </option>
    <option
      value={16}
    >
      16
    </option>
    <option
      value={17}
    >
      17
    </option>
    <option
      value={18}
    >
      18
    </option>
    <option
      value={19}
    >
      19
    </option>
    <option
      value={20}
    >
      20
    </option>
    <option
      value={21}
    >
      21
    </option>
    <option
      value={22}
    >
      22
    </option>
    <option
      value={23}
    >
      23
    </option>
    <option
      value={24}
    >
      24
    </option>
    <option
      value={25}
    >
      25
    </option>
    <option
      value={26}
    >
      26
    </option>
    <option
      value={27}
    >
      27
    </option>
    <option
      value={28}
    >
      28
    </option>
    <option
      value={29}
    >
      29
    </option>
    <option
      value={30}
    >
      30
    </option>
    <option
      value={31}
    >
      31
    </option>
  </select>
  <select
    className="dc-month"
    disabled={false}
    name="selectMonth"
    onBlur={[Function]}
    onChange={[Function]}
    value="Invalid date"
  >
    <option
      value=""
    >
      Bulan
    </option>
    <option
      value="01"
    >
      Januari
    </option>
    <option
      value="02"
    >
      Februari
    </option>
    <option
      value="03"
    >
      Maret
    </option>
    <option
      value="04"
    >
      April
    </option>
    <option
      value="05"
    >
      Mei
    </option>
    <option
      value="06"
    >
      Juni
    </option>
    <option
      value="07"
    >
      Juli
    </option>
    <option
      value="08"
    >
      Agustus
    </option>
    <option
      value="09"
    >
      September
    </option>
    <option
      value={10}
    >
      Oktober
    </option>
    <option
      value={11}
    >
      November
    </option>
    <option
      value={12}
    >
      Desember
    </option>
  </select>
  <select
    className="dc-year"
    disabled={false}
    name="selectYear"
    onBlur={[Function]}
    onChange={[Function]}
    value="Invalid date"
  >
    <option
      value=""
    >
      Tahun
    </option>
    <option
      value={2019}
    >
      2019
    </option>
  </select>
</div>
`;
